### ISSUE TITLE
Sync failing on imperfect internet

### ISSUE THREAD

**Issue by woj-tek** [ankidroid/Anki-Android#4498]
###### Research
*Enter an [ x ] character to confirm the points below:*

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

###### Reproduction Steps

1. Open ankidroid
2. try to synchronise using spotty internet connection

###### Expected Result
Ankidroid syncronises correctly

###### Actual Result
Error about "connection timing out" occurs frequently.
On the same internet connection I can easily sunchronise Anki desktop.

###### Debug info
AnkiDroid Version = 2.7

Android Version = 6.0.1

ACRA UUID = b4c70967-9c6a-460d-a312-bb847734398b

**Comment by woj-tek** [comment]
So it looks like it's caused by timeout - would it be possible to increase it (or make it somehow configurable)? I've tried digging a bit in the sources (not very familiar with Android) and couldn't find where/how to do it.

- error
`<string name="sync_connection_error">La conexión ha expirado. O bien hay problemas con su conexión a internet  o bien el fichero es demasiado grande.</string>`
- com/ichi2/anki/DeckPicker.java:1492
`                    } else if (resultType.equals("connectionError")) {
                        dialogMessage = res.getString(R.string.sync_connection_error);
                        showSyncErrorMessage(joinSyncMessages(dialogMessage, syncMessage));
`
- com/ichi2/async/Connection.java:404
`                } catch (RuntimeException e) {
                    if (timeoutOccured(e)) {
                        data.result = new Object[] {"connectionError" };
`

From my observation the timeout is around 1minute, which with following connection may be a bit problematic:

> Ping Type: ICMP
> Packet Size: default
> TTL: default
> Timeout: 3000
> Count: 30
> Interval: 1000
> General Timeout: not used
> 
> PING ankiweb.net (148.251.54.24)
> connection timeout 1
> connection timeout 2
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=3 ttl=44 time=249.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=4 ttl=44 time=246.0 ms
> connection timeout 5
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=6 ttl=44 time=247.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=7 ttl=44 time=247.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=8 ttl=44 time=246.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=9 ttl=44 time=246.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=10 ttl=44 time=245.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=11 ttl=44 time=246.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=12 ttl=44 time=246.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=13 ttl=45 time=249.0 ms
> connection timeout 14
> connection timeout 15
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=16 ttl=44 time=253.0 ms
> connection timeout 17
> connection timeout 18
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=19 ttl=45 time=247.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=20 ttl=44 time=250.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=21 ttl=44 time=271.0 ms
> connection timeout 22
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=23 ttl=44 time=247.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=24 ttl=44 time=246.0 ms
> connection timeout 25
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=26 ttl=44 time=247.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=27 ttl=44 time=246.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=28 ttl=44 time=247.0 ms
> 64 bytes from ankisrs.net (148.251.54.24): icmp_seq=29 ttl=44 time=246.0 ms
> connection timeout 30
> 
> 
> --- ping statistics ---
> 30 packets transmitted, 20 received, 33% packet loss, time 69480ms
> rtt min/avg/max/mdev = 245/248/271/5,7 ms

And:
> traceroute to ankiweb.net (148.251.54.24), 30 hops max
> Hop 1:
>     From 10.50.1.113, 39 ms
> 
> Hop 2:
>     From 10.50.1.26, 40 ms
> 
> Hop 3:
>     From grtvapem2-xe6-1-11-0.net.telefonicaglobalsolutions.com (216.184.112.244), 46 ms
> 
> Hop 4:
>     From ae6-grtmiabr1.net.telefonicaglobalsolutions.com (94.142.121.190), 137 ms
> 
> Hop 5:
>     From xe10-1-10-0-grtwaseq2.net.telefonicaglobalsolutions.com (5.53.5.45), 164 ms
> 
> Hop 6:
>     From xe1-1-3-0-grtfraix5.net.telefonicaglobalsolutions.com (94.142.118.161), 266 ms
> 
> Hop 7:
>     From 213.140.49.121, 258 ms
> 
> Hop 8:
>     From decix2-gw.hetzner.de (80.81.193.164), 251 ms
> 
> Hop 9:
>     From decix2-gw.hetzner.de (80.81.193.164), 260 ms
> 
> Hop 10:
>     From ex9k2.rz20.hetzner.de (213.239.229.10), 274 ms
> 
> Hop 11:
>     From ankisrs.net (148.251.54.24), 268 ms
> 
> Traceroute complete: 11 hops, time: 19480 ms

**Comment by timrae** [comment]
I don't mind increasing from 30s to 60s... I suppose during this time you're not able to cancel the sync though...? So that the only option is to force-kill AnkiDroid...

**Comment by timrae** [comment]
Can you submit a PR and we can discuss it there?

**Comment by woj-tek** [comment]
Hi, sorry for the lack of the response, I've been occupied with other stuff and in the meantime I ended up with better connection so the issue has gone away.
To answer you question - I was able to cancel the synchronisation during that time (though, it still looks like it was waiting for the time-out to kick-in instead of being immediate). However cancelling the synchronisation is not a solution if your end goal is having data in sync.
